### ISSUE TITLE
DNS Update automation

### ISSUE THREAD

**Issue by bartsmykla** [kubernetes/k8s.io#783]
As I started working on automating the process of running octodns whenever [DNS configuration](https://github.com/kubernetes/k8s.io/tree/master/dns/zone-configs) will change here is my step by step plan:

1. [Done: #805, #822, [k/test-infra#17449](https://github.com/kubernetes/test-infra/pull/17449), #827, #828] ~Building consensus around if we want our [octodns container image](/dns/octodns-docker) to be build and managed using [our own image promotion process](/k8s.gcr.io)~
   - ~[If yes]~:
      ~1. [Creating staging repository](/k8s.gcr.io#creating-staging-repos)~
      ~2. [Enabling automatic build](/k8s.gcr.io#enabling-automatic-builds)~
   - ~[if not]: Figure out which image you should use to proceed further~

2. [Done: #949, Closed: #862] ~Modifying current [`dns/push.sh`](https://github.com/kubernetes/k8s.io/blob/master/dns/push.sh) script to be able to run update without docker involved and modifying `make` - ref. [@spiffxp comment](https://github.com/kubernetes/k8s.io/issues/783#issuecomment-618007941)~

3. [Done: #865] ~Ensuring the service account which will be used to run `octodns` exists and have enough permissions to do it's job~

   ~Initially proposed (adding `empower_dns_updater` logic to [lib.sh](/kubernetes/k8s.io/blob/master/infra/gcp/lib.sh), and call the function in [ensure-main-project.sh](/kubernetes/k8s.io/blob/master/infra/gcp/ensure-main-project.sh)):~
   ```bash
   # Ensure the dns_updater service account exists and has the ability to update DNSes.
   # $1: DNS Updater Service Account name
   # $2: The GCP project
   function empower_dns_updater() {
       if [ ! $# -eq 2 -o -z "$1" -o -z "$2" ]; then
           echo "empower_dns_updater(service_account_name, project) requires 2 arguments" >&2
           return 1
       fi
       local service_account_name="$1"
       local project="$2"
       local service_account=$(svc_acct_email "$project" "$service_account_name")

       ensure_service_account "$project" "$service_account_name" "k8s-infra dns updater"

       # Allow dns_updater to adminsiter DNSes.
       gcloud projects add-iam-policy-binding "$project" \
           --member "serviceAccount:$service_account" \
           --role roles/dns.admin
   }
   ```

   ```bash
   PROJECT="kubernetes-public"
   DNS_UPDATOR_SVCACCT_NAME="k8s-infra-dns-updater"

   empower_dns_updater "$DNS_UPDATOR_SVCACCT_NAME" "$PROJECT"
   ```

   ~After the discussion at one of the last community calls we decided it's ok to instead of explicitly grant permissions for service account to add it to proper google groups with existing permissions (`k8s-infra-dns-admins@kubernetes.io`) so instead of creating the function as it was planned before, we can add to [ensure-main-project.sh](/kubernetes/k8s.io/blob/master/infra/gcp/ensure-main-project.sh):~

   ```bash
   color 6 "Ensuring the k8s-infra-dns-updater serviceaccount exists"
   ensure_service_account \
     "${PROJECT}" \
     "k8s-infra-dns-updater" \
     "k8s-infra dns updater"
   ```

4. [Done: #865] ~Empowering the Prow KSA to authenticate as service account from steps above (adding logic to [ensure-main-project.sh](infra/gcp/ensure-main-project.sh), as suggested by @listx - ref. [#issuecomment-617929312](https://github.com/kubernetes/k8s.io/issues/783#issuecomment-617929312)~

   ```bash
   PROJECT="kubernetes-public"

   color 6 "Empowering k8s-infra-dns-updater serviceaccount to be used on build cluster"
   empower_ksa_to_svcacct \
       "k8s-prow.svc.id.goog[test-pods/k8s-infra-dns-updater]" \
       "${PROJECT}" \
       $(svc_acct_email "${PROJECT}" "k8s-infra-dns-updater")
   ```

5. [Done: #920, Closed: [k/test-infra#17739](https://github.com/kubernetes/test-infra/pull/17739)] ~Adding created service account to [`k/test-infra/config/prow/cluster/trusted_serviceaccounts.yaml`](https://github.com/kubernetes/test-infra/blob/bc6a077f2ca569a84fdca22807021c419165212e/config/prow/cluster/trusted_serviceaccounts.yaml)~

   ```yaml
   ---
   kind: ServiceAccount
   apiVersion: v1
   metadata:
     annotations:
       iam.gke.io/gcp-service-account: k8s-infra-dns-updater@kubernetes-public.iam.gserviceaccount.com
     name: k8s-infra-dns-updater
     namespace: test-pods
   ```

6. [Done: [`k/test-infra#17940`](https://github.com/kubernetes/test-infra/pull/17940)] ~Creating `postsubmit` prow job [here](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/wg-k8s-infra)~:

   ```yaml
   postsubmits:
     kubernetes/k8s.io:
     - name: k8sio-dns-update
       cluster: k8s-infra-prow-build-trusted
       decorate: true
       run_if_changed: "^dns/zone-configs/"
       max_concurrency: 1
       branches:
       - ^master$
       annotations:
         testgrid-create-test-group: 'true'
         testgrid-dashboards: wg-k8s-infra-k8sio
         testgrid-alert-email: k8s-infra-alerts@kubernetes.io
         testgrid-num-failures-to-alert: '1'
       spec:
         serviceAccountName: k8s-infra-dns-updater
         containers:
         - image: us.gcr.io/k8s-artifacts-prod/infra-tools/octodns:v20200612-d4db040
           command:
           - bash
           args:
           - -c
           - "cd dns && ./push.sh --dry-run"
   ```
7. ~I'm not sure who should run this script, but as far as I understand someone who has access to prow cluster should run [`k/test-infra/workload-identity/bind-service-accounts.sh`](https://github.com/kubernetes/test-infra/blob/master/workload-identity/bind-service-accounts.sh) (@listx please correct me if I'm wrong)~ [There is no need to do that - ref. [#issuecomment-617929312](https://github.com/kubernetes/k8s.io/issues/783#issuecomment-617929312)]

8. ~Updating `providers.config.directory` in [`octodns-config.yaml`](/dns/octodns-config.yaml) to reflect clone path which the container will be decoreted with (ref. [`prow/pod-utilities.md#how-to-configure`](https://github.com/kubernetes/test-infra/blob/master/prow/pod-utilities.md#how-to-configure) which in our case would be `/home/prow/go/src/github.com/kubernetes/k8s.io/dns/zone-configs`~ (as during the [`dns/push.sh` refactoring](https://github.com/kubernetes/k8s.io/pull/862/files#diff-079d3dc93d7bdf142659677856d874baR166-R206), there was implemented logic to precook the config file and it doesn't matter what is set in `providers.config.directory` I think it's unnecessary to change it)

9. Figuring out how to add alerts for this job which should send messages to our [#k8s-infra-alerts slack channel](https://kubernetes.slack.com/archives/CUD6HL48N) (cc. @thockin)
10. Testing for some time if everything works as expected
11. Building consensus if we can switch the job from `dry-run` to `do-it` mode
12. Switching the mode at our prow job

@dims @thockin @spiffxp what do you think about it?

**Comment by listx** [comment]
> As I started working on automating the process of running octodns whenever [DNS configuration](https://github.com/kubernetes/k8s.io/tree/master/dns/zone-configs) will change here is my step by step plan:
> 
> 1. Building consensus around if we want our [octodns container image](/dns/octodns-docker) to be build and managed using [our own image promotion process](/k8s.gcr.io)
>    
>    * [If yes]:
>      
>      1. [Creating staging repository](/k8s.gcr.io#creating-staging-repos)
>      2. [Enabling automatic build](/k8s.gcr.io#enabling-automatic-builds)
>    * [if no]: Figure out which image you should use to proceed further
> 2. Creating service account which will be used to run `octodns`
> 3. Adding created service account to [`k/test-infra/prow/cluster/trusted_serviceaccounts.yaml`](https://github.com/kubernetes/test-infra/blob/master/prow/cluster/trusted_serviceaccounts.yaml)
> 4. I'm not sure who should run this script, but as far as I understand someone who has access to prow cluster should run [`k/test-infra/workload-identity/bind-service-accounts.sh`](https://github.com/kubernetes/test-infra/blob/master/workload-identity/bind-service-accounts.sh) (@listx please correct me if I'm wrong)

The `bind-service-accounts.sh` script does not need to be run manually; last time I added Workload Identity stuff, the only things in k/test-infra that had to be changed was (1) the `trusted_serviceaccounts.yaml` file which you already noted and (2) the prow job definition itself. Creation of the Kubernetes Service Account (KSA) on Prow and making that KSA try to authenticate as the GCP account from Step 2 is done automatically by test-infra.

That being said, this step needs to be replaced with something like https://github.com/kubernetes/k8s.io/pull/718, where from our GCP project's perspective, we allow the KSA to authenticate as our GCP SA.

> 5. Updating `providers.config.directory` in [`octodns-config.yaml`](/dns/octodns-config.yaml) to reflect clone path which the container will be decoreted with (ref. [`prow/pod-utilities.md#how-to-configure`](https://github.com/kubernetes/test-infra/blob/master/prow/pod-utilities.md#how-to-configure) which in our case would be `/home/prow/go/src/github.com/kubernetes/k8s.io/dns/zone-configs`
> 6. Creating `postsubmit` prow job [here](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/wg-k8s-infra):
> 
> ```yaml
> postsubmits:
>   kubernetes/k8s.io:
>   - name: k8sio-dns-update
>     annotations:
>       testgrid-dashboards: wg-k8s-infra-k8sio
>     decorate: true
>     run_if_changed: "^dns/zone-configs/"
>     max_concurrency: 1
>     branches:
>     - ^master$
>     spec:
>       containers:
>       - image: [to_be_defined]
>         command:
>         - octodns-sync
>         args:
>         - --config-file
>         - /home/prow/go/src/github.com/kubernetes/k8s.io/dns/octodns-config.yaml
>         - --log-stream-stdout
>         - --debug
> ```

This is missing the `serviceAccountName: <KSA_NAME>` field under `spec`.

**Comment by spiffxp** [comment]
> 1. Building consensus around if we want our [octodns container image](/dns/octodns-docker) to be build and managed using [our own image promotion process](/k8s.gcr.io)

yes

> 2. Creating service account which will be used to run `octodns`
> 4. I'm not sure who should run this script, but as far as I understand someone who has access to prow cluster should run [`k/test-infra/workload-identity/bind-service-accounts.sh`](https://github.com/kubernetes/test-infra/blob/master/workload-identity/bind-service-accounts.sh) (@listx please correct me if I'm wrong)

these should be done in this repo via the `infra/gcp/ensure-*.sh` scripts (not sure offhand if it slots in existing, or should be a new script)

>3. Adding created service account to [`k/test-infra/prow/cluster/trusted_serviceaccounts.yaml`](https://github.com/kubernetes/test-infra/blob/master/prow/cluster/trusted_serviceaccounts.yaml)
>5. Updating `providers.config.directory` in [`octodns-config.yaml`](/dns/octodns-config.yaml) to reflect clone path which the container will be decoreted with (ref. [`prow/pod-utilities.md#how-to-configure`](https://github.com/kubernetes/test-infra/blob/master/prow/pod-utilities.md#how-to-configure) which in our case would be `/home/prow/go/src/github.com/kubernetes/k8s.io/dns/zone-configs`
>6. Creating `postsubmit` prow job [here](https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/wg-k8s-infra):
```yaml
postsubmits:
  kubernetes/k8s.io:
  - name: k8sio-dns-update
    annotations:
      testgrid-dashboards: wg-k8s-infra-k8sio
    decorate: true
    run_if_changed: "^dns/zone-configs/"
    max_concurrency: 1
    branches:
    - ^master$
    spec:
      containers:
      - image: [to_be_defined]
        command:
        - octodns-sync
        args:
        - --config-file
        - /home/prow/go/src/github.com/kubernetes/k8s.io/dns/octodns-config.yaml
        - --log-stream-stdout
        - --debug
```
These can all be done in the same PR to test-infra.  As @listx said, this needs a `serviceAccountName: foo` to use the foo service account via workload identity.

Also, please don't hardcode a prow path into the config file.  I'm pretty sure the working directory in prow is the root of first repo cloned, so it should be fine to reference `./dns/octodns-config` and `./dns-zone-configs`.  

Is a direct `octodns-sync` really what we want to use? I would prefer we use the same entrypoints locally and in CI, whether that's `make` or `./dns/push.sh` etc.

>7. Figuring out how to add alerts for this job which should send messages to our [#k8s-infra-alerts slack channel](https://kubernetes.slack.com/archives/CUD6HL48N) (cc. @thockin)

This should use testgrid to email an alert to k8s-infra-alerts@kubernetes.io if the job fails, e.g.: https://github.com/kubernetes/test-infra/blob/ca5fb3795b846243453d7b403221860cb8d668b5/config/jobs/kubernetes/test-infra/test-infra-trusted.yaml#L725-L728

TIL we have a #k8s-infra-alerts.  I'm less inclined to use it since I use e-mail for job failure alerts, but if you want to try, you should be able to setup something specific for kubernetes/k8s.io jobs here: https://github.com/kubernetes/test-infra/blob/ca5fb3795b846243453d7b403221860cb8d668b5/config/prow/config.yaml#L76-L87 (unclear if you can specify which _specific_ jobs vs. job types)

> 8. Testing for some time if everything works as expected
> 9. Building consensus if we can switch the job from `dry-run` to `do-it` mode
> 10. Switching the mode at our prow job

no objections here

**Comment by bartsmykla** [comment]
@listx :

> The bind-service-accounts.sh script does not need to be run manually; last time I added Workload Identity stuff, the only things in k/test-infra that had to be changed was (1) the trusted_serviceaccounts.yaml file which you already noted and (2) the prow job definition itself. Creation of the Kubernetes Service Account (KSA) on Prow and making that KSA try to authenticate as the GCP account from Step 2 is done automatically by test-infra.

Thank you for that comments, it's much clearer in my head now!

> That being said, this step needs to be replaced with something like #718, where from our GCP project's perspective, we allow the KSA to authenticate as our GCP SA.

addressed

> This is missing the serviceAccountName: <KSA_NAME> field under spec.

addressed

@spiffxp :

> these should be done in this repo via the `infra/gcp/ensure-*.sh` scripts (not sure offhand if it slots in existing, or should be a new script)

Indeed. Updated steps to reflect that

> These can all be done in the same PR to test-infra. As @listx said, this needs a serviceAccountName: foo to use the foo service account via workload identity.

addressed

> Also, please don't hardcode a prow path into the config file. I'm pretty sure the working directory in prow is the root of first repo cloned, so it should be fine to reference ./dns/octodns-config and ./dns-zone-configs.

addressed

> Is a direct octodns-sync really what we want to use? I would prefer we use the same entrypoints locally and in CI, whether that's make or ./dns/push.sh etc.

addressed (ref. step no. 2) and also addressed these changes at step no. 8

**Comment by spiffxp** [comment]
agreed upon name of repo to host images, bart to move forward on the plan described above

**Comment by bartsmykla** [comment]
As #805 was merged I opened #822 and [k/test-infra#17449](https://github.com/kubernetes/test-infra/pull/17449) to automate building and pushing of images.

The [`ensure-staging-storage.sh`](https://github.com/kubernetes/k8s.io/blob/master/infra/gcp/ensure-staging-storage.sh) script still needs to be run by @cblecker, @dims, @spiffxp or @thockin (ref. [#805#issuecomment-621706127](https://github.com/kubernetes/k8s.io/pull/805#issuecomment-621706127)) before the [k/test-infra#17449](https://github.com/kubernetes/test-infra/pull/17449) will be merged.

**Comment by bartsmykla** [comment]
- As octodns image is now build and stored in registry managed by us (ref. [us.gcr.io/k8s-artifacts-prod/infra-tools/octodns](https://us.gcr.io/k8s-artifacts-prod/infra-tools/octodns) I updated the issue.
- TIL you can force `make` to pass the arguments to it's targets by using:
  ```make
  ifeq (run, $(firstword $(MAKECMDGOALS)))
    runargs := $(wordlist 2, $(words $(MAKECMDGOALS)), $(MAKECMDGOALS))
    $(eval $(runargs):;@true)
  endif
  ```
  and then passing `$(runargs)` to the target directive, I changed a little bit proposed postsubmit job template's spec

**Comment by bartsmykla** [comment]
Created PR with the refactoring of [`dns/push.sh`]https://github.com/kubernetes/k8s.io/blob/8689b98013f88cb63559457373a003ca2e8080ba/dns/push.sh() (ref. #862)
cc. @spiffxp @thockin

**Comment by bartsmykla** [comment]
As there were changes with some of the plans I updated the first issue with the state of the efforts and correct order

**Comment by bartsmykla** [comment]
As #865 was merged and service account exists I updated the issue

**Comment by bartsmykla** [comment]
Also created [k/test-infra#17739](https://github.com/kubernetes/test-infra/pull/17739) for step no. 5 about marking service account: `k8s-infra-dns-updator` as a trusted

**Comment by bartsmykla** [comment]
Let's give me some update about this issue. As I decided to close #862 after comments form @thockin and @spiffxp I opened new PR (#949), with the only changes related to removing docker from the [`dns/push.sh`](https://github.com/kubernetes/k8s.io/blob/9644f6a360d7b26c79b1ab543e3c00987550cbf4/dns/push.sh) and [`dns/check-zone.sh`](https://github.com/kubernetes/k8s.io/blob/9644f6a360d7b26c79b1ab543e3c00987550cbf4/dns/check-zone.sh) which was merged, then the new version of octodns image was built (ref. [Prow Job: `post-k8sio-infra-tools-push-images`](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/post-k8sio-infra-tools-push-images/1271581607601377282) and [`GCP Console/GCR: k8s-staging-infra-tools`](https://console.cloud.google.com/gcr/images/k8s-staging-infra-tools/GLOBAL/octodns@sha256:f74d145d124712fe3f75e751f6e5f6d0403b0c0b265dae825927a8ff4e84c547/details?tab=info&project=k8s-staging-infra-tools&authuser=0&organizationId=758905017065)) and then promoted (ref. #953, [GCP Console/GCR: k8s-artifacts-prod](https://console.cloud.google.com/gcr/images/k8s-artifacts-prod/US/infra-tools/octodns@sha256:f74d145d124712fe3f75e751f6e5f6d0403b0c0b265dae825927a8ff4e84c547/details?tab=info&project=k8s-artifacts-prod&authuser=0&organizationId=758905017065) and I updated [`dns/Makefile`](https://github.com/kubernetes/k8s.io/blob/9644f6a360d7b26c79b1ab543e3c00987550cbf4/dns/Makefile#L22) (ref. #954) to use the newly promoted image.

The next step is to create actual Prow Job.

**Comment by bartsmykla** [comment]
I have updated the first issue with all changes arising from the last post.

Also opened PR with adding the prow job [`k/test-infra#17940`](https://github.com/kubernetes/test-infra/pull/17940)

**Comment by spiffxp** [comment]
https://github.com/kubernetes/test-infra/pull/17940 has been merged so I think we are now waiting on declaration of victory to call this done (e.g. have we seen any jobs successfully deploy DNS changes)

**Comment by spiffxp** [comment]
Should also update the README for dns to reference the automation instead of humans doing things (though leaving instructions on how humans can manually do things is still helpful, e.g. https://github.com/kubernetes/k8s.io/blob/master/groups/README.md)

**Comment by bartsmykla** [comment]
@spiffxp as everything looks fine, and the last jobs handled by the prow were successful I think we can close the issue and declare a victory. :-)

I have also opened the PR with documentation changes related to automation: https://github.com/kubernetes/k8s.io/pull/1222

**Comment by bartsmykla** [comment]
As the PR with dock was merged  think we can close the issue.

/close
